design-docs/Defer draft API.md

```diff
@@ -0,0 +1,852 @@
+# Apollo Kotlin: @defer draft API
+
+This document outlines proposed changes to the user-facing API of Apollo Kotlin related to supporting the new `@defer`
+directive.
+
+Implementation is not discussed for now, the idea is to first agree on the desired API, other documents will dive into
+implementation.
+
+>
+> 💡 Note: even though both `@defer` and `@stream` are part of the same spec, it has been decided to focus on `@defer`
+> first, as it appears `@stream` has a lower demand.
+>
+
+## Context
+
+- Proposed GraphQL spec changes: https://github.com/graphql/graphql-spec/pull/742
+- Original RFC: https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md
+- Working Group: https://github.com/robrichard/defer-stream-wg
+
+This directive applies to fragment spreads and inline fragments. When present, the fragment’s fields can be received
+incrementally.
+
+For example:
+
+```graphql
+type Query {
+  computers: [Computer!]!
+}
+
+type Computer {
+  id: ID!
+  cpu: String!
+  year: Int!
+  screen: Screen!
+}
+
+type Screen {
+  resolution: String!
+  isColor: Boolean!
+}
+```
+
+```graphql
+query Query {
+  computers {
+    id
+    ... on Computer @defer {
+      cpu
+      year
+      screen {
+        resolution
+        ... on Screen @defer {
+          isColor
+        }
+      }
+    }
+  }
+}
+```
+
+Would be received as:
+
+```json
+// Payload 1
+{
+  "data": {
+    "computers": [
+      {
+        "id": "Computer1"
+      },
+      {
+        "id": "Computer2"
+      }
+    ]
+  },
+  "hasNext": true
+}
+
+// Payload 2
+{
+  "data": {
+    "__typename": "Computer",
```

Comment:
Is this __typename automatically sent with all defered payload? This looks weird this it wasn't requested in the query.

Reply:
Good catch! No, they're not automatically sent, my mistake!